cloudstack/service_offering_constrained_resource.go

```diff
@@ -0,0 +1,270 @@
+	}
+	if !plan.ServiceOfferingDiskQosStorage.IsNull() {
+		resp.Diagnostics.Append(plan.ServiceOfferingDiskQosStorage.As(ctx, &planDiskQosStorage, basetypes.ObjectAsOptions{})...)
+	}
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// common params
+	params := r.client.ServiceOffering.NewCreateServiceOfferingParams(plan.DisplayText.ValueString(), plan.Name.ValueString())
+	plan.commonCreateParams(ctx, params)
+	planDiskQosHypervisor.commonCreateParams(ctx, params)
+	planDiskOffering.commonCreateParams(ctx, params)
+	planDiskQosStorage.commonCreateParams(ctx, params)
+
+	// resource specific params
+	if !plan.CpuSpeed.IsNull() {
+		params.SetCpuspeed(int(plan.CpuSpeed.ValueInt32()))
+	}
+	if !plan.MaxCpuNumber.IsNull() {
+		params.SetMaxcpunumber(int(plan.MaxCpuNumber.ValueInt32()))
+	}
+	if !plan.MaxMemory.IsNull() {
+		params.SetMaxmemory(int(plan.MaxMemory.ValueInt32()))
+	}
+	if !plan.MinCpuNumber.IsNull() {
+		params.SetMincpunumber(int(plan.MinCpuNumber.ValueInt32()))
+	}
+	if !plan.MinMemory.IsNull() {
+		params.SetMinmemory(int(plan.MinMemory.ValueInt32()))
+	}
+
+	// create offering
+	cs, err := r.client.ServiceOffering.CreateServiceOffering(params)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error creating service offering",
+			"Could not create constrained offering, unexpected error: "+err.Error(),
+		)
+		return
+	}
+
+	//
+	plan.Id = types.StringValue(cs.Id)
+
+	//
+	resp.Diagnostics.Append(resp.State.Set(ctx, plan)...)
+}
+
+func (r *serviceOfferingConstrainedResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	//
+	var state serviceOfferingConstrainedResourceModel
+	var stateDiskQosHypervisor ServiceOfferingDiskQosHypervisor
+	var stateDiskOffering ServiceOfferingDiskOffering
+	var stateDiskQosStorage ServiceOfferingDiskQosStorage
+
+	//
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if !state.ServiceOfferingDiskQosHypervisor.IsNull() {
+		resp.Diagnostics.Append(state.ServiceOfferingDiskQosHypervisor.As(ctx, &stateDiskQosHypervisor, basetypes.ObjectAsOptions{})...)
+	}
+	if !state.ServiceOfferingDiskOffering.IsNull() {
+		resp.Diagnostics.Append(state.ServiceOfferingDiskOffering.As(ctx, &stateDiskOffering, basetypes.ObjectAsOptions{})...)
+	}
+	if !state.ServiceOfferingDiskQosStorage.IsNull() {
+		resp.Diagnostics.Append(state.ServiceOfferingDiskQosStorage.As(ctx, &stateDiskQosStorage, basetypes.ObjectAsOptions{})...)
+	}
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	//
+	cs, _, err := r.client.ServiceOffering.GetServiceOfferingByID(state.Id.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error reading service offering",
+			"Could not read constrained service offering, unexpected error: "+err.Error(),
+		)
+		return
+	}
+
+	// resource specific
+	if cs.Cpuspeed > 0 {
+		state.CpuSpeed = types.Int32Value(int32(cs.Cpuspeed))
+	}
+	// These fields arent returned from list
+	// max_cpu_number
```

Comment:
You should be able to get these values from the details in the response.

Reply:
You're right.  Added these into the state.